Building libcxx with KLEE.

rsprabery opened the pull request:
More work is needed to get things like re2 to run. The `klee_driver.cpp`
needs to be aware of size limits.

@laurentsimon PTAL.

jonathanmetzman commented on fuzzers/klee/builder.Dockerfile:
```diff
@@ -49,38 +49,53 @@ RUN git clone https://github.com/stp/stp.git /stp && \
-    git checkout debug && \
+    git checkout debug
+
+# the libcxx build script in the KLEE repo depends on wllvm:
```
nit: capitalize first letter in comments please.

jonathanmetzman commented on fuzzers/klee/fuzzer.py:
```diff
@@ -38,7 +38,8 @@ def prepare_build_environment():
     os.environ['CXX'] = 'gclang++'
 
-    os.environ['FUZZER_LIB'] = '/libAFL.a -L/ -lKleeMock'
+    os.environ['FUZZER_LIB'] = '/libAFL.a -L/ -lKleeMock -lpthread'
```
I think `-pthread` is set by default in CXXFLAGS and CFLAGS.
Is it actually needed?

rsprabery replied:
It was needed to resolve an issue with libpng, though the rest of that fix is missing in this PR.

jonathanmetzman commented on fuzzers/klee/fuzzer.py:
```diff
@@ -66,13 +67,23 @@ def emptydir(path):
+    if ulimit_cmd:
+        ulimit_command = [ulimit_cmd + ';']
+        ulimit_command.extend(command)
+        ret = subprocess.call(' '.join(ulimit_command),
```
Don't think the join call is needed. `call` should work on a list.

rsprabery replied:
I double checked, in general, call works with lists as you said - but I'm getting an issue where the side effects of ulimit -s aren't visible to klee (increased stack size) unless I do this join.

jonathanmetzman commented on fuzzers/klee/fuzzer.py:
```diff
@@ -66,13 +67,23 @@ def emptydir(path):
 
-def run(command, hide_output=False):
-    """Run a the command |command|"""
+def run(command, hide_output=False, ulimit_cmd=None):
```
Why is this the ulimit stuff needed?

kren1 replied:
STP uses recursion when constraint solving. If you have a stack limit KLEE reaches it quite quickly and dies. However I'm not sure if these work inside docker, usually you have to tell docker, that the container has no stack limit.

rsprabery replied:
As @kren1 mentioned, this is necessary for STP. 

I've tested this with both libpng and re2 (fixes for those specific benchmarks coming in another PR) and this fixed the issue.

I had set the unlimit in the dockerfile, but it didn't appear to take. This was the only way I was able to clear up the issue with STP. Open to suggestions if anyone has a cleaner way to do this.

laurentsimon commented:
FYI, AFL also needs this, but does it itself in afl-fuzz.c
<https://github.com/google/AFL/blob/master/afl-fuzz.c#L2023> I don't think
KLEE has a similar option, does it?

>

rsprabery commented:
Thanks for the feedback! I've addressed the comments. I think the ulimit thing I've got going here is a temporary fix for now. We'll want to do something to make that cleaner. Maybe add an option to KLEE itself?

jonathanmetzman commented:
Didn't bother checking that presubmit wasn't one of the failures. Could you fix the please? (no rush)
